src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -228,9 +230,110 @@ public static char ToCharLower(int value)
+                Vector128<ushort> vec1 = Vector128.LoadUnsafe(ref srcRef, offset);
+                Vector128<ushort> vec2 = Vector128.LoadUnsafe(ref srcRef, offset + (nuint)Vector128<ushort>.Count);
+                Vector128<byte> vec = Vector128.Narrow(vec1, vec2);
+
+                // Based on "Algorithm #3" https://github.com/WojciechMula/toys/blob/master/simd-parse-hex/geoff_algorithm.cpp
+                // by Geoff Langdale and Wojciech Mula
+                // Move digits '0'..'9' into range 0xf6..0xff.
+                Vector128<byte> t1 = vec + Vector128.Create((byte)(0xFF - '9'));
+                // And then correct the range to 0xf0..0xf9.
+                // All other bytes become less than 0xf0.
+                Vector128<byte> t2 = SubtractSaturate(t1, Vector128.Create((byte)6));
+                // Convert into uppercase 'a'..'f' => 'A'..'F' and
+                // move hex letter 'A'..'F' into range 0..5.
+                Vector128<byte> t3 = (vec & Vector128.Create((byte)0xDF)) - Vector128.Create((byte)'A');
+                // And correct the range into 10..15.
+                // The non-hex letters bytes become greater than 0x0f.
+                Vector128<byte> t4 = AddSaturate(t3, Vector128.Create((byte)10));
+                // Convert '0'..'9' into nibbles 0..9. Non-digit bytes become
+                // greater than 0x0f. Finally choose the result: either valid nibble (0..9/10..15)
+                // or some byte greater than 0x0f.
+                Vector128<byte> nibbles = Vector128.Min(t2 - Vector128.Create((byte)0xF0), t4);
+                // Any high bit is a sign that input is not a valid hex data
+                if (!Utf16Utility.AllCharsInVector128AreAscii(vec1 | vec2) ||
+                    AddSaturate(nibbles, Vector128.Create((byte)(127 - 15))).ExtractMostSignificantBits() != 0)
+                {
+                    // Input is either non-ASCII or invalid hex data
+                    break;
+                }
+                Vector128<byte> output;
+                if (Ssse3.IsSupported)
+                {
+                    output = Ssse3.MultiplyAddAdjacent(nibbles,
+                        Vector128.Create((short)0x0110).AsSByte()).AsByte();
+                }
+                else
+                {
+                    // Workaround for missing MultiplyAddAdjacent on ARM
+                    Vector128<short> ones = Vector128.Create(0x10010).AsInt16();
+                    Vector128<short> vl = AdvSimd.Multiply(
+                        AdvSimd.ZeroExtendWideningLower(nibbles.GetLower()).AsInt16(), ones).AsInt16();
+                    Vector128<short> vu = AdvSimd.Multiply(
+                        AdvSimd.ZeroExtendWideningLower(nibbles.GetUpper()).AsInt16(), ones).AsInt16();
+                    output = AdvSimd.AddSaturate(
+                        AdvSimd.Arm64.UnzipEven(vl, vu),
+                        AdvSimd.Arm64.UnzipOdd(vl, vu)).AsByte();
```

Comment:
Is there not a better way to do this either? Arm64 has many specialized multiply instructions, including those that simultaneously `widen` or do `addition` or `saturation`.

@TamarChristinaArm might have some insight...

--------

For reference, `Vector128<short> MultiplyAddAdjacent(Vector128<byte> left, Vector128<sbyte> right)` is `PMADDUBSW`. This effectively does the following:
```
DEST[0] = SaturateToSignedWord((short)(SRC[1] * DEST[1]) + (short)(SRC[0] * DEST[0]));
DEST[1] = SaturateToSignedWord((short)(SRC[3] * DEST[3]) + (short)(SRC[2] * DEST[2]));
...
```

That is, it takes an `unsigned byte` (`left`) and a `signed byte` (`right`). It multiplies these to produce a `signed short`. It then does a pairwise addition.

---------------

In the case of xarch, we're taking an `unsigned byte` (`nibbles`) of unknown value and a `signed byte` of known value `0x0110` which is effectively unsigned.

Since for `x-bit * x-bit = y-bit` the sign is only relevant when `y-bit > x-bit`, we can effectively treat this as an `unsigned multiplication` and then reinterpret it back as `signed`.

That is, we can at least do:
```
Vector128<byte> multiplier = Vector128.Create((ushort)0x0110).AsByte());
Vector128<short> vl = AdvSimd.MultiplyWideningLower(nibbles.GetLower(), multiplier.GetLower()).AsInt16();
Vector128<short> vu = AdvSimd.MultiplyWideningUpper(nibbles, multiplier).AsInt16()
```

This will allow us to remove a couple of the operations. I'm not aware of a trivial way to do a pairwise saturation here, however so we may need to keep that latter bit as is.

Reply:
What I did here is I took a general impl of `MultiplyAddAdjacent` and simplified it for the constant argument. I ran benchmarks on M1 and they overall demonstrated similar ratios as for x64. I also do think it's possible to make it more efficient

Reply:
Hi :),

I'm on holidays till Monday, but Tanner's version is much better.
Unfortunately AArch64 lacks a lot of pairwise operations.

However from a quick look (so please double check the below before implementing) looking at the constant you used

```
Vector128<short> ones = Vector128.Create(0x10010).AsInt16();
```

This looks like it's multiplying even elements by `0x10` and odd by `0x1`.
So evens are just shift left by 4 and odds just zero extend?

Atm the sequence from Tanner is:

```
movi + umull + umull + (uzp1 + uzp2) + sqadd
```

where the permutes run in parallel. So roughly 2 + 4 + 4 + 2 + 2 = 14 cycles on Neoverse-N1, 12 cycles assuming the `movi` got lifted.

Since you need the permute anyway and `nibbles` is unsigned, how about instead, create a zero vector outside the loop and do the permutes earlier to separate even/odd, but take advantage of this to get a zero extend free.

```
zip1 (with zero vector) + uzp2 + ushll 4 + sqadd
```

Which would be 2 + 2 + 2 = 6 since the permutes again run in parallel.

Should be a 2x speedup.

Reply:
@TamarChristinaArm @tannergooding  thanks for the suggestion! Did I get it correctly:
```csharp
var zip1 = AdvSimd.Arm64.ZipLow(nibbles, Vector128<byte>.Zero);
var uzp2 = AdvSimd.Arm64.UnzipOdd(nibbles, nibbles);
var ushl = AdvSimd.ShiftLeftLogicalWideningLower(uzp2.GetLower(), 4);
output = AdvSimd.AddSaturate(zip1, ushl.AsByte());
```
?
unfortunately it doesn't produce expected results 🙁

Reply:
Unsure about the `uzip2`, but I think the principle is that because the multiplier is effectively multiplying `even * 16` and `odd * 1` and that we know both inputs are unsigned, we can do this:
```
var even = AdvSimd.Arm64.ZipLow(nibbles, Vector128<byte>.Zero).AsInt16();
var odd  = AdvSimd.Arm64.ZipHigh(nibbles, Vector128<byte>.Zero).AsInt16();
even = AdvSimd.ShiftLeftLogicalWideningLower(even, 4);
output = AdvSimd.AddSaturate(even, odd);
```

This gives us a 4 instruction sequence: `zip1, zip2, ushll, sqadd`

-------------

This works because `ZipLow` and `ZipHigh` interleave the elements. That is `ZipLow` is:
```
for (int p = 0; p < Vector128<T>.Count / 2; p++)
{
    result[(2 * p) + 0] = left[p];
    result[(2 * p) + 1] = right[p];
}
```

While `ZipHigh` is:
```
for (int p = 0; p < Vector128<T>.Count / 2; p++)
{
    result[(2 * p) + 0] = left[p + 1];
    result[(2 * p) + 1] = right[p + 1];
}
```

This means we can get the effect of decoupling the even and odd elements while zero-extending at the same time.

We then just need to multiply the even results by 16, which is the same as shifting left by 4. We don't need to touch the odd results since they are just multiplied by 1.

Then we do the saturating add and are complete.

Reply:
I think you can also do `var odd  = AdvSimd.Arm64.ZipHigh(nibbles, even).AsInt16();` as well. That should ensure that we don't get two `movni reg, #0` emitted for `Vector128<T>.Zero`...

That should "work" since the odd bytes of `ZipLow` will be zero

Reply:
Ah, that's not quite right 

`ZipLow` is doing `lhs[0], rhs[0], lhs[1], rhs[1], ...`
and we want `lhs[0], rhs[0], lhs[2], rhs[2], ...`

Reply:
This should work.. TransposeEven is the one that does `lhs[0], rhs[0], lhs[2], rhs[2], ...` (And odd indices for TransposeOdd)

```csharp
var even = AdvSimd.TransposeEven(nibbles, Vector128<byte>.Zero).AsInt16();
var odd  = AdvSimd.TransposeOdd(nibbles, even.AsByte()).AsInt16();
even = AdvSimd.ShiftLeftLogical(even, 4);
output = AdvSimd.AddSaturate(even, odd);
```

Reply:
> This should work.. TransposeEven is the one that does lhs[0], rhs[0], lhs[2], rhs[2], ...

Args. yes sorry, flipped the permutes operation in my head. So should be:
```c#
var zip1 = AdvSimd.Arm64.TransposeEven(nibbles, Vector128<byte>.Zero); // trn1
var uzp2 = AdvSimd.Arm64.UnzipOdd(nibbles, nibbles); // uzp2
var ushl = AdvSimd.ShiftLeftLogicalWideningLower(uzp2.GetLower(), 4); // ushll 4
output = AdvSimd.AddSaturate(zip1, ushl.AsByte()); // sqadd
```
or @tannergooding 's version above.

> we don't get two movni reg, #0 emitted for Vector128<T>.Zero

The loop doesn't look that big, and I assume there's no loop invariant motion? It might be worth lifting the 5 constants out of the loop.  You'll definitely have enough left over on Advanced SIMD, but am unsure about AVX/SSE. But worth trying to see what comes out.

Reply:
> I assume there's no loop invariant motion? 

We do loop invariant motion, it's just never done for simple things like zero vectors. Even if I do hoisting by hands it will be propagated back at its use. The main issue that our RA is not perfect so we try to minimize cheap cases to occupy registers. We'll improve that eventually. Overall, most constant vectors are hoisted from this loop by JIT (at least till the IsAscii check)